compiler/parser/parser.peg

```diff
@@ -992,7 +992,10 @@ UnaryPlusOrMinus
 PlusOrMinusOp = ("+" / "-") { return string(c.text), nil }
 
 DerefExpr
-  = expr:DerefExpr "[" __ from:AdditiveExpr __ ":" __ to:AdditiveExpr? __ "]" {
+  = expr:DerefExpr __ "::" __ typ:(TypeAsValue / TypeLiteral) {
+      return &ast.Cast{Kind: "Cast", Expr: expr.(ast.Expr), Type: typ.(ast.Expr), Loc: loc(c)}, nil
```

Comment:
We've been going with the pattern that different syntaxes should be described using a different syntax node and here we are borrowing the node used for function call-like casts (e.g., `string(1)`).